// omm/src/serializer/Serializer.ts
import ObjectRetriever from "./ObjectRetriever"
import Document from "./Document"
import SubObjectPath from "./SubObjectPath"
import {TypeClass, PersistenceAnnotation, getId, setNonEnumerableProperty, className } from "../annotations/PersistenceAnnotation"

export default class Serializer {
    private objectRetriever:ObjectRetriever;

    constructor(retri:ObjectRetriever){
        this.objectRetriever = retri;
    }

    static init(){
        PersistenceAnnotation.getEntityClasses().forEach(function (c:TypeClass<Object>) {
            Serializer.installLazyLoaderGetterSetters(c);
        });
    }

    private static installLazyLoaderGetterSetters(c:TypeClass<Object>){
        PersistenceAnnotation.getTypedPropertyNames(c).forEach(function (propertyName:string) {
            if (PersistenceAnnotation.isStoredAsForeignKeys(c, propertyName)) {
                //console.log("On Class " + className + ": creating lazy loader for " + propertyName);
                var propertyDescriptor = Object.getOwnPropertyDescriptor(c.prototype, propertyName);

                Object.defineProperty(c.prototype, propertyName, {
                    get: function ():any {
                        // TODO this doesnt work for subdocuments
                        //console.log("Monkey patched getter "+propertyName);
                        var v:any;
                        if (propertyDescriptor && propertyDescriptor.get)
                            v = propertyDescriptor.get.apply(this);
                        else
                            v = this["_" + propertyName];
                        if (Serializer.needsLazyLoading(this, propertyName)) {
                            var objectRetriever:ObjectRetriever = this["_objectRetriever"];
                            if (typeof v == "string") {
                                //console.log("Lazy loading " + className + "." + propertyName);
                                v = objectRetriever.getObject(v, this, propertyName );
                                this[propertyName] = v;
                            }
                            else  // TODO this could be improved so that it loads them when they are accessed rather than to load them all at once
                            {
                                //console.log("Lazy loading array/map " + className + "." + propertyName);
                                for( var i in v )
                                {
                                    var ele = v[i];
                                    v[i] = objectRetriever.getObject(ele, this, propertyName );
                                }
                            }
                        }
                        //console.log("Monkey patched getter "+propertyName+" returns ",v);
                        return v;
                    },
                    set: function (v:any) {
                        //console.log("Monkey patched setter " + propertyName + " v:" + v);

                        if (propertyDescriptor && propertyDescriptor.set)
                            propertyDescriptor.set.apply(this, arguments);
                        else {
                            if (!Object.getOwnPropertyDescriptor(this, "_" + propertyName)) {
                                Object.defineProperty(this, "_" + propertyName, {
                                    configurable: false,
                                    enumerable: false,
                                    writable: true
                                });
                            }
                            this["_" + propertyName] = v;
                        }
                    },
                    configurable: propertyDescriptor ? propertyDescriptor.configurable : true,
                    enumerable: propertyDescriptor ? propertyDescriptor.enumerable : true
                });
            }
            else{
                //console.log("On Class " + className + ": no lazy loader for " + propertyName);
            }

        });
    }

    static forEachTypedObject(object:Object, cb:(path:SubObjectPath, object:Object)=>void ){
        this.forEachTypedObjectRecursive(object,object, new SubObjectPath(), [], cb);

    }

    static  forEachTypedObjectRecursive(rootObject:Object, object:Object, path:SubObjectPath, visited:Array<Object>,cb:(path:SubObjectPath, object:Object)=>void):void {
        var that = this;
        if (visited.indexOf(object) != -1)
            return;
        if (!object || typeof object != "object")
            return;
        visited.push(object);

        cb(path, object);
        var objectClass = PersistenceAnnotation.getClass(object);
        PersistenceAnnotation.getTypedPropertyNames(objectClass).forEach(function (typedPropertyName:string) {
            if (!PersistenceAnnotation.isStoredAsForeignKeys(objectClass, typedPropertyName)&&!PersistenceAnnotation.isParent(objectClass, typedPropertyName)) {
                //console.log("updating foreignkey property " + typedPropertyName);
                var v:Object = object[typedPropertyName];
                if (v) {
                    if (PersistenceAnnotation.isArrayOrMap(objectClass, typedPropertyName)) {
                        //console.log("updating foreignkey property " + typedPropertyName + " is array");
                        for (var i in v) {
                            var e = v[i];
                            //console.log("updating persistnece path for isArrayOrMap " + typedPropertyName + "  key:" + i + " value:", e, "object: ", object);
                            var subObjectPath = path.clone();
                            if (typeof getId(e)!="undefined") {
                                subObjectPath.appendArrayOrMapLookup(typedPropertyName, getId(e));
                            } else {
                                subObjectPath.appendArrayOrMapLookup(typedPropertyName, i);
                            }
                            cb(subObjectPath, e);
                            that.forEachTypedObjectRecursive(rootObject, e, subObjectPath, visited, cb);
                        }
                    }
                    else {
                        var subObjectPath = path.clone();
                        subObjectPath.appendPropertyLookup(typedPropertyName);
                        cb(subObjectPath, v);
                        that.forEachTypedObjectRecursive(rootObject, v, subObjectPath, visited, cb);
                    }
                }

            }
            //else {
            //    //console.log( "foreign key "+typedPropertyName );
            //    if (!Serializer.needsLazyLoading(object, typedPropertyName)) {
            //        var v:MeteorPersistable = object[typedPropertyName];
            //        if (v) {
            //            if (PersistenceAnnotation.isArrayOrMap(objectClass, typedPropertyName)) {
            //                for (var i in v) {
            //                    var e = v[i];
            //                    var subObjectPath = path.clone();
            //                    if (e.getId && e.getId()) {
            //                        subObjectPath.appendArrayOrMapLookup(typedPropertyName, e.getId());
            //                    } else {
            //                        subObjectPath.appendArrayOrMapLookup(typedPropertyName, i);
            //                    }
            //                    cb(subObjectPath, v);
            //                }
            //            }
            //            else if (!v._serializationPath)
            //                that.updateSerializationPaths(v, visited);
            //        }
            //    }else{
            //
            //    }
            //}
        });
    }


    static needsLazyLoading(object:Object, propertyName:string) {
        // TODO inheritance
        var oc = PersistenceAnnotation.getClass(object);
        if( PersistenceAnnotation.isStoredAsForeignKeys(oc, propertyName ) )
        {
            var shadowpropertyDescriptor = Object.getOwnPropertyDescriptor(object, "_" + propertyName);
            var shadowPropertyIsKeys = false;
            if (shadowpropertyDescriptor)
                if (typeof object["_" + propertyName] == "string")
                    shadowPropertyIsKeys = true;
                else if (PersistenceAnnotation.isArrayOrMap(oc, propertyName)) {
                    var v = object["_" + propertyName];
                    for( var i in v )
                    {
                        if(typeof v[i] =="string" )
                            shadowPropertyIsKeys = true;
                        break;

                    }
                }
            return shadowPropertyIsKeys;
        }
        else
            return false;

    }

    toObject<T extends Object>(doc:Document, f?:TypeClass<T>):T {
        var o:T;
        if(Array.isArray(doc)){
            var r = [];
            for( var j=0; j<(<Array<any>>doc).length; j++ ){
                r[j] = this.toObjectRecursive(doc[j], undefined, f);
            }
            o = <any>r;
        } else if ( !doc || typeof doc == "string" || typeof doc == "number"  || typeof doc == "date" || typeof doc == "boolean")
            o =  <T>doc;
        else
            o =  this.toObjectRecursive(doc,undefined, f);
        this.objectRetriever.postToObject(o);
        return o;
    }

    private toObjectRecursive<T extends Object>(doc:Document, parent:Object, f?:TypeClass<T>):T {
        var o:T;
        if( !doc )
            return <T>doc;
        if(typeof doc=="function")
            throw new Error("Error in 'toObject'. doc is a function.");



        if ( f && typeof f["toObject"]=="function" ) {
                //console.log("using the custom toObject function of class "+omm.className(f));
                o = f["toObject"](doc);
        } else {
            // if the document contains a property called "className" it defines the class that's going to be instantiated
            if (doc.className){
                f = PersistenceAnnotation.getEntityClassByName(doc.className);
            }
            if(!f)
                throw new Error("Could not determine class of document. Either the document needs to have a 'className' property or a class needs to be passed to the serializer. Document: "+ JSON.stringify( doc ) );
            // instantiate the new object
            o = new f();

            // iterate over all properties
            for (var propertyName in doc) {
                if( propertyName=="className" )
                    continue;
                var value = doc[propertyName];
                var objectNameOfTheProperty:string = PersistenceAnnotation.getObjectPropertyName(f, propertyName);
                if(!objectNameOfTheProperty)
                    objectNameOfTheProperty = propertyName;
                var propertyClass = PersistenceAnnotation.getPropertyClass(f, objectNameOfTheProperty);
                var isStoredAsKeys = PersistenceAnnotation.isStoredAsForeignKeys(f, objectNameOfTheProperty);

                if (propertyClass && !isStoredAsKeys) {
                    if (PersistenceAnnotation.isArrayOrMap(f, objectNameOfTheProperty)) {
                        var result = Array.isArray(value) ? [] : {};
                        for (var i in value) {
                            var entry:Document = value[i];
                            entry = this.toObjectRecursive(entry, o, propertyClass);
                            result[i] = entry;
                        }
                        // this can only happen once because if the property is accessed the "lazy load" already kicks in
                        o[objectNameOfTheProperty] = result;
                    }
                    else {
                        o[objectNameOfTheProperty] = this.toObjectRecursive(value, o, propertyClass);
                    }
                }
                else {
                    o[objectNameOfTheProperty] = value;
                }
            }
            PersistenceAnnotation.getParentPropertyNames(f).forEach(function (parentPropertyName:string) {
                o[parentPropertyName] = parent;
            });
        }
        setNonEnumerableProperty(o, "_objectRetriever", this.objectRetriever);
        //o._objectRetriever = this.objectRetriever;
        return o;
    }

    toDocument(object:Object ):Document {
        this.objectRetriever.preToDocument( object );
        return this.toDocumentRecursive(object);
    }

    private toDocumentRecursive(object:any, rootClass?:TypeClass<Object>, parentObject?:Object, propertyNameOnParentObject?:string):Document {
        var result:Document;
        if ( !object || typeof object == "string" || typeof object == "number"  || typeof object == "date" || typeof object == "boolean")
            result =  <Document>object;
        else if( Array.isArray(object) ){
            result = [];
            for( var i=0; i<object.length; i++ ){
                result[i] = this.toDocumentRecursive(object[i]);
            }
        } else {
            var objectClass =  PersistenceAnnotation.getClass(object);
            if( typeof (<any>objectClass).toDocument == "function" ){
                result = (<any>objectClass).toDocument( object );
            } else {
                var parentClass = PersistenceAnnotation.getClass(parentObject);
                if (parentObject && propertyNameOnParentObject && PersistenceAnnotation.isStoredAsForeignKeys(parentClass, propertyNameOnParentObject)) {
                    return <Document><any>this.objectRetriever.getId(object);
                }
                else {
                    result = this.createDocument(object, rootClass ? rootClass : PersistenceAnnotation.getClass(object), parentObject, propertyNameOnParentObject);

                    // if the class of the object does not correspond to the expected type, we add it to the document
                    if( parentClass && objectClass!=PersistenceAnnotation.getPropertyClass(parentClass, propertyNameOnParentObject))
                        result.className = className(objectClass);
                }
            }
        }
        //console.log("returning document:",result);
        return result;
    }

    private createDocument(object:any, rootClass?:TypeClass<Object>, parentObject?:Object, propertyNameOnParentObject?:string):Document {
        var doc:any = {};
        var objectClass = PersistenceAnnotation.getClass(object);
        for (var property in object) {
            var value = object[property];
            var documentNameOfTheProperty:string = PersistenceAnnotation.getDocumentPropertyName(objectClass,property);
            if( !documentNameOfTheProperty )
                documentNameOfTheProperty = property;
            if (value !== undefined && !PersistenceAnnotation.isIgnored(objectClass, property)&& !PersistenceAnnotation.isParent(objectClass, property)) {
                // primitives
                if( PersistenceAnnotation.getPropertyClass(objectClass,property) ) {

                    // array
                    if (PersistenceAnnotation.isArrayOrMap(objectClass, property)) {
                        var result;
                        if (Array.isArray(value))
                            result = [];
                        else
                            result = {};

                        for (var i in value) {
                            var subObject = value[i];
                            result[i] = this.toDocumentRecursive(subObject, rootClass, object, property);
                        }
                        doc[documentNameOfTheProperty] = result;
                    }
                    // object
                    else if (typeof object[property] == 'object') {
                        doc[documentNameOfTheProperty] = this.toDocumentRecursive(value, rootClass, object, property);
                    } else {
                        throw new Error("Unsupported type : "+ typeof value);
                    }
                }
                else if (typeof value == 'function') {
                    // not doing eeeeenithing with functions
                }
                else {
                    doc[documentNameOfTheProperty] = value;
                }

            }

        }
        return <Document>doc;
    }
}
