// omm/spec/classes/TestTree.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
///<reference path="./references.d.ts"/>
var omm = require("../../src/omm");
var Tests = require("./Tests");
var TestTree = (function () {
    function TestTree(initialHeight) {
        this.height = 10;
        this.leaves = [];
        this.height = initialHeight || 10;
    }
    TestTree.prototype.grow = function () {
        this.height++;
        //console.log("Tree is growing to new heights: ", this.height+" on the "+(omm.getMeteor().isServer?"server":"client"));
        this.leaves.push(new Tests.TestLeaf("leaf" + this.getHeight(), this));
        this.leaves.forEach(function (l) {
            l.grow();
        });
        if (omm.getMeteor().isClient)
            return "grown on the client";
        else
            return "grown on the server";
    };
    TestTree.prototype.wither = function () {
        this.leaves = [];
        omm.emit("gardenevents", "withered");
        omm.emit("gardenevents", "withered2");
    };
    TestTree.prototype.thisThrowsAnError = function () {
        throw new Error("Hello world");
    };
    TestTree.prototype.getHeight = function () {
        return this.height;
    };
    TestTree.prototype.getLeaves = function () {
        return this.leaves;
    };
    TestTree.prototype.growAndReturnLeaves = function () {
        this.grow();
        return this.leaves;
    };
    __decorate([
        omm.Id, 
        __metadata('design:type', String)
    ], TestTree.prototype, "treeId", void 0);
    __decorate([
        omm.ArrayType("TestLeaf"),
        omm.DocumentName('thoseGreenThings'), 
        __metadata('design:type', Array)
    ], TestTree.prototype, "leaves", void 0);
    __decorate([
        omm.Type("TestAddress"),
        omm.AsForeignKey, 
        __metadata('design:type', Tests.TestAddress)
    ], TestTree.prototype, "address", void 0);
    __decorate([
        omm.Wrap, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', []), 
        __metadata('design:returntype', String)
    ], TestTree.prototype, "grow", null);
    __decorate([
        omm.Wrap, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', []), 
        __metadata('design:returntype', void 0)
    ], TestTree.prototype, "wither", null);
    __decorate([
        omm.Wrap, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', []), 
        __metadata('design:returntype', void 0)
    ], TestTree.prototype, "thisThrowsAnError", null);
    __decorate([
        omm.CollectionUpdate,
        omm.MeteorMethod({ replaceWithCall: true, resultType: "TestLeaf" }), 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', []), 
        __metadata('design:returntype', void 0)
    ], TestTree.prototype, "growAndReturnLeaves", null);
    TestTree = __decorate([
        omm.Entity, 
        __metadata('design:paramtypes', [Number])
    ], TestTree);
    return TestTree;
}());
exports.TestTree = TestTree;
