// omm/spec/classes/TestPerson.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
///<reference path="./references.d.ts"/>
var omm = require("../../src/omm");
var Tests = require("./Tests");
var TestPerson = (function () {
    function TestPerson(id, name) {
        this.addresses = [];
        this.trees = [];
        this.phoneBook = {};
        this.wood = {};
        this.family = {};
        this.gardenState = 10;
        this._id = id;
        this.name = name;
    }
    TestPerson.prototype.getId = function () {
        return this._id;
    };
    TestPerson.prototype.setId = function (id) {
        this._id = id;
    };
    //rename(n:string):void
    //{
    //    this.name = n;
    //}
    TestPerson.prototype.getName = function () {
        return this.name;
    };
    TestPerson.prototype.getAddresses = function () {
        return this.addresses;
    };
    TestPerson.prototype.getTree = function () {
        return this.tree;
    };
    TestPerson.prototype.collectLeaf = function () {
        //console.log("collecting leaf:",this.tree);
        this.leaf = this.tree.getLeaves()[0];
    };
    TestPerson.prototype.chooseTree = function (t) {
        //console.log("choosing tree:",t);
        this.tree = t;
    };
    TestPerson.prototype.chooseLeaf = function (l) {
        this.leaf = l;
    };
    TestPerson.prototype.rename = function (n) {
        this.name = n;
        return this.name;
    };
    TestPerson.prototype.collectionUpdateRename = function (n) {
        this.name = "Collection Update:" + n;
        return this.name;
    };
    TestPerson.prototype.addAddress = function (a) {
        //console.log("inside add address:", (a instanceof Tests.TestAddress));
        this.addresses.push(a);
        return a;
    };
    TestPerson.prototype.addAddresses = function (addresses) {
        var that = this;
        addresses.forEach(function (a) {
            //console.log('street:',a.getStreet());
            that.addresses.push(a);
        });
        return this.addresses;
    };
    TestPerson.prototype.fromServer = function (cb) {
        if (omm.getMeteor().isClient)
            throw new Error("though shall not be called on the client");
        cb(undefined, omm.getMeteor().isServer);
    };
    TestPerson.prototype.addToWood = function (t, s) {
        this.trees.push(t);
        if (s)
            this.wood[s] = t;
    };
    TestPerson.prototype.addFamilyRelation = function (s, p) {
        if (s)
            this.family[s] = p;
    };
    TestPerson.prototype.addPhoneNumber = function (s, p) {
        this.phoneBook[s] = p;
    };
    TestPerson.prototype.tendToGarden = function () {
        this.gardenState++;
        this.tree.grow();
        return this.tree.getHeight();
    };
    __decorate([
        omm.Type("TestPhoneNumber"), 
        __metadata('design:type', Tests.TestPhoneNumber)
    ], TestPerson.prototype, "phoneNumber", void 0);
    __decorate([
        omm.ArrayOrMap("TestAddress"), 
        __metadata('design:type', Array)
    ], TestPerson.prototype, "addresses", void 0);
    __decorate([
        omm.Type("TestTree"),
        omm.AsForeignKeys, 
        __metadata('design:type', Tests.TestTree)
    ], TestPerson.prototype, "tree", void 0);
    __decorate([
        omm.Type("TestLeaf"),
        omm.AsForeignKeys, 
        __metadata('design:type', Tests.TestLeaf)
    ], TestPerson.prototype, "leaf", void 0);
    __decorate([
        omm.ArrayOrMap("TestLeaf"),
        omm.AsForeignKeys, 
        __metadata('design:type', Array)
    ], TestPerson.prototype, "trees", void 0);
    __decorate([
        omm.ArrayOrMap("TestPhoneNumber"), 
        __metadata('design:type', Object)
    ], TestPerson.prototype, "phoneBook", void 0);
    __decorate([
        omm.ArrayOrMap("TestTree"),
        omm.AsForeignKeys, 
        __metadata('design:type', Object)
    ], TestPerson.prototype, "wood", void 0);
    __decorate([
        omm.ArrayOrMap("TestPerson"),
        omm.AsForeignKeys, 
        __metadata('design:type', Object)
    ], TestPerson.prototype, "family", void 0);
    __decorate([
        omm.Wrap, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', []), 
        __metadata('design:returntype', void 0)
    ], TestPerson.prototype, "collectLeaf", null);
    __decorate([
        omm.Wrap, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Tests.TestTree]), 
        __metadata('design:returntype', void 0)
    ], TestPerson.prototype, "chooseTree", null);
    __decorate([
        omm.Wrap, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Tests.TestLeaf]), 
        __metadata('design:returntype', void 0)
    ], TestPerson.prototype, "chooseLeaf", null);
    __decorate([
        omm.CollectionUpdate,
        omm.MeteorMethod, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [String]), 
        __metadata('design:returntype', String)
    ], TestPerson.prototype, "rename", null);
    __decorate([
        omm.CollectionUpdate,
        omm.MeteorMethod, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [String]), 
        __metadata('design:returntype', String)
    ], TestPerson.prototype, "collectionUpdateRename", null);
    __decorate([
        omm.CollectionUpdate,
        omm.MeteorMethod({ parameterTypes: ["TestAddress"], replaceWithCall: true }), 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Tests.TestAddress]), 
        __metadata('design:returntype', Tests.TestAddress)
    ], TestPerson.prototype, "addAddress", null);
    __decorate([
        omm.CollectionUpdate,
        omm.MeteorMethod({ parameterTypes: ["TestAddress"], replaceWithCall: true }), 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Array]), 
        __metadata('design:returntype', Array)
    ], TestPerson.prototype, "addAddresses", null);
    __decorate([
        omm.MeteorMethod({ parameterTypes: ["callback"], replaceWithCall: true, serverOnly: true }), 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Function]), 
        __metadata('design:returntype', void 0)
    ], TestPerson.prototype, "fromServer", null);
    __decorate([
        omm.Wrap, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Tests.TestTree, String]), 
        __metadata('design:returntype', void 0)
    ], TestPerson.prototype, "addToWood", null);
    __decorate([
        omm.Wrap, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [String, Tests.TestPerson]), 
        __metadata('design:returntype', void 0)
    ], TestPerson.prototype, "addFamilyRelation", null);
    __decorate([
        omm.Wrap, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [String, Tests.TestPhoneNumber]), 
        __metadata('design:returntype', void 0)
    ], TestPerson.prototype, "addPhoneNumber", null);
    __decorate([
        omm.Wrap, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', []), 
        __metadata('design:returntype', Number)
    ], TestPerson.prototype, "tendToGarden", null);
    TestPerson = __decorate([
        omm.Entity, 
        __metadata('design:paramtypes', [String, String])
    ], TestPerson);
    return TestPerson;
}());
exports.TestPerson = TestPerson;
