// omm/spec/classes/TestPersonCollection.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var omm = require("../../src/omm");
var Tests = require("./Tests");
var TestPersonCollection = (function (_super) {
    __extends(TestPersonCollection, _super);
    function TestPersonCollection() {
        _super.call(this, Tests.TestPerson);
    }
    TestPersonCollection.prototype.newPerson = function (n, callback) {
        var p = new Tests.TestPerson();
        p.name = n;
        var that = this;
        this.insert(p, function (e, id) {
            callback(e, id ? that.getById(id) : undefined);
        });
    };
    TestPersonCollection.prototype.insertPerson = function (n) {
        var p = new Tests.TestPerson();
        p.name = n;
        var that = this;
        var id = this.insert(p);
        return this.getById(id);
    };
    TestPersonCollection.staticInsertPerson = function (n) {
        var personCollection = omm.getRegisteredObject("TestPersonCollection");
        var p = new Tests.TestPerson();
        p.name = n;
        var id = personCollection.insert(p);
        return personCollection.getById(id);
    };
    TestPersonCollection.staticInsertPerson2 = function (n) {
        var personCollection = omm.getRegisteredObject("TestPersonCollection");
        var p = new Tests.TestPerson();
        p.name = n;
        var id = personCollection.insert(p);
        return personCollection.getById(id);
    };
    TestPersonCollection.prototype.haveBaby = function (mom, dad, callback) {
        //console.log("mom: ", mom);
        //console.log("dad: ", dad);
        var kid = new Tests.TestPerson();
        kid.name = "child of " + mom.name + " and " + dad.name;
        kid.family["mom"] = mom;
        kid.family["dad"] = dad;
        var that = this;
        this.insert(kid, function (e, id) {
            //console.log("The baby is inserted into the database");
            callback(e, id ? that.getById(id) : undefined);
        });
    };
    TestPersonCollection.prototype.removePerson = function (id, callback) {
        this.remove(id, callback);
    };
    TestPersonCollection.prototype.removeAllPersons = function (callback) {
        var _this = this;
        this.find({}).forEach(function (testPerson) {
            _this.remove(testPerson.getId());
        });
        callback();
    };
    TestPersonCollection.init = function () {
        omm.getMeteor().startup(function () {
            if (omm.getMeteor().isServer) {
                omm.getMeteor().publish("persons", function () {
                    return omm.Collection.getByName("TestPerson").getMeteorCollection().find({});
                });
            }
            else {
                omm.getMeteor().subscribe("persons");
            }
            omm.registerObject('TestPersonCollection', new Tests.TestPersonCollection());
        });
    };
    __decorate([
        omm.MeteorMethod({ object: 'TestPersonCollection', replaceWithCall: true, parameterTypes: ["string", "callback"] })
    ], TestPersonCollection.prototype, "newPerson", null);
    __decorate([
        omm.MeteorMethod({ object: 'TestPersonCollection' })
    ], TestPersonCollection.prototype, "insertPerson", null);
    __decorate([
        omm.MeteorMethod({ object: 'TestPersonCollection', replaceWithCall: true, parameterTypes: ["TestPerson", "TestPerson", "callback"] })
    ], TestPersonCollection.prototype, "haveBaby", null);
    __decorate([
        omm.MeteorMethod({ object: 'TestPersonCollection', replaceWithCall: true, parameterTypes: ["string", "callback"] })
    ], TestPersonCollection.prototype, "removePerson", null);
    __decorate([
        omm.MeteorMethod({ object: 'TestPersonCollection', replaceWithCall: true, parameterTypes: ["callback"] })
    ], TestPersonCollection.prototype, "removeAllPersons", null);
    __decorate([
        omm.StaticMeteorMethod
    ], TestPersonCollection, "staticInsertPerson", null);
    __decorate([
        omm.StaticMeteorMethod('helloWorld', { parameterTypes: ['string'] })
    ], TestPersonCollection, "staticInsertPerson2", null);
    return TestPersonCollection;
}(omm.Collection));
exports.TestPersonCollection = TestPersonCollection;
