// omm/spec/classes/TestLeaf.js
/**
 * Created by bert on 07.05.15.
 */
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var omm = require("../../src/omm");
var Tests = require("./Tests");
var TestLeaf = (function () {
    function TestLeaf(id, parent) {
        this._id = id;
        this.greenNess = 1;
        this.parent = parent;
    }
    TestLeaf.prototype.getId = function () {
        return this._id;
    };
    TestLeaf.prototype.grow = function () {
        this.greenNess++;
    };
    TestLeaf.prototype.getTree = function () {
        return this.parent;
    };
    TestLeaf.prototype.flutter = function () {
        this.greenNess++;
        omm.emit("fluttering");
    };
    __decorate([
        omm.DocumentName("greenIndex"), 
        __metadata('design:type', Number)
    ], TestLeaf.prototype, "greenNess", void 0);
    __decorate([
        omm.Parent, 
        __metadata('design:type', Tests.TestTree)
    ], TestLeaf.prototype, "parent", void 0);
    __decorate([
        omm.Wrap, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', []), 
        __metadata('design:returntype', void 0)
    ], TestLeaf.prototype, "flutter", null);
    TestLeaf = __decorate([
        omm.Entity, 
        __metadata('design:paramtypes', [String, Tests.TestTree])
    ], TestLeaf);
    return TestLeaf;
}());
exports.TestLeaf = TestLeaf;
