// omm/src/omm/Config.ts
/**
 * Created by bert on 22.03.16.
 */

export interface MeteorInterface {
    Error? : {
        new( error: string | number, reason?: string, details?: string ): any;
    };
    users? : MongoCollectionInterface;
    wrapAsync? : (func:Function, context?:Object) => any;
    call?: (method:string, ...parameters:any[]) => void;
    methods? : (IMeteorMethodsDictionary:any) => any;
    startup? : ( a:()=>void ) => void;
    isServer? : boolean;
    isClient? : boolean;
    publish?: (name:string, f:( args:any ) => MongoCursorInterface)=>void;
    subscribe?: ( name:string ) => void;
}

export interface ObjectIDStatic {
    new(hexString?: string): ObjectID;
}
export interface ObjectID {
    _str?:string;
}

export interface MongoInterface {
    Collection?:{
        new( name:string ):MongoCollectionInterface;
    };
    ObjectID: ObjectIDStatic;
}

export interface MongoCollectionInterface {
    find(search:any):MongoCursorInterface;
    insert( obj:any );
    remove( id:string );
    update( pattern:any, data:any );
}

export interface MongoCursorInterface {
    fetch():any[];
}


import {environmentReferences} from "../annotations/PersistenceAnnotation"

export function config( options:{ Meteor? : MeteorInterface, Mongo?: MongoInterface } ) {
    if( options.Meteor )
        environmentReferences.meteorReference = options.Meteor;
    else
        environmentReferences.meteorReference = <any>Meteor;

    if( options.Mongo ) {
        environmentReferences.mongoReference = options.Mongo;
    } else{
        environmentReferences.mongoReference = <any>Mongo;
    }
}

export function getMeteor():MeteorInterface{
    return environmentReferences.meteorReference;
}

export function getMongo():MongoInterface{
    return environmentReferences.mongoReference;
}
